java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/TreeItem.java

```diff
@@ -56,6 +55,27 @@
     public String resourceUri;
     // ?: string | MarkdownString | undefined
     public Object tooltip;
+    
+    // NBLS-specific items, to be processed by the client:
+    // Metadata for the icon.
+    public IconDescriptor iconDescriptor;
+    // The icon's index
+    public int iconIndex;
+    
+    /**
+     * Metadata that describe an icon origin or contents.
+     */
+    public static class IconDescriptor {
+        /**
+         * Base URI / imageId of the icon.
+         */
+        public URI baseUri;
+        
+        /**
+         * Supplemental IDs from composed merged-in images or applied filters.
+         */
+        public String[] composition;
```

Comment:
As discussed earlier, trying to replace icons after they have already been loaded and even merged with badges etc. seems like a big hack. But you know the requirements here better than me, and it's localized to the LSP server stuff, so it's fine with me.

Reply:
The point is that (after some further adjustments), the client could get just icon's metadata (base URL + composition info) - and then ***decide*** to either handle the picture itself - or ask for the resource's bits back to the server: in that case, the original image's bits will be served.
This may seem awkward, but I don't think a single LSP server can satisfy various client needs / UI styles

Reply:
Yeah, I still don't quite understand what the LSP server _is_. A screenshot would still be appreciated!